### PR TITLE
[BEAM-7981] Fix double iterable stripping

### PR DESCRIPTION
**Pull request by udim**
... in CallableWrapperDoFn

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by udim** [comment]
python :docs failure should be fixed in https://github.com/apache/beam/pull/9714

**Comment by udim** [comment]
R: @robertwb @aaltay

File: sdks/python/apache_beam/transforms/core.py [hunk 1]
```diff
@@ -585,7 +585,7 @@ def default_type_hints(self):
       try:
         fn_type_hints.strip_iterable()
       except ValueError as e:
-        raise ValueError('Return value not iterable: %s: %s' % (self, e))
+        logging.warning('%s: %s', self.default_label(), e)
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: It's unclear to me under what conditions it's OK for this not to be iterable.

File: sdks/python/apache_beam/transforms/core.py [hunk 2]
```diff
@@ -676,22 +676,17 @@ def __repr__(self):
...
-    # TODO(robertwb): Should we require an iterable specification for FlatMap?
-    if type_hints.output_types:
-      args, kwargs = type_hints.output_types
-      if len(args) == 1 and is_consistent_with(
-          args[0], typehints.Iterable[typehints.Any]):
-        type_hints = type_hints.copy()
-        type_hints.set_output_types(element_type(args[0]), **kwargs)
+    # Do not modify self._fn's type hints object. This method may be called more
+    # than once (such as once for input and once for output type checks) and
+    # strip_iterable modifies the object.
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/core.py]: Oh :(. Can we please make strip_iterable return the result rather than have it mutate self?

File: sdks/python/apache_beam/typehints/typed_pipeline_test_py3.py [hunk 3]
```diff
@@ -107,14 +105,16 @@ class MyDoFn(beam.DoFn):
       def process(self, element: int) -> str:
         return str(element)
 
-    with self.assertRaisesRegex(ValueError, r'str.*is not iterable'):
+    with self.assertLogs() as cm:
```
Comment (robertwb) [inline sdks/python/apache_beam/typehints/typed_pipeline_test_py3.py]: Why this change? (Similarly below.)

File: sdks/python/apache_beam/transforms/core.py (hunk 1 above)
Comment (udim) [reply sdks/python/apache_beam/transforms/core.py]: I automatically made this a warning to be consistent with the warning for the other invocation of strip_iterable below.
Below, the code historically allowed output type hints of DoFns to not be iterable. I would like it to be an exception when the output type is not iterable, but it seems that it would break existing pipelines (at least in the case below).

File: sdks/python/apache_beam/transforms/core.py (hunk 2 above)
Comment (udim) [reply sdks/python/apache_beam/transforms/core.py]: done

File: sdks/python/apache_beam/typehints/typed_pipeline_test_py3.py (hunk 3 above)
Comment (udim) [reply sdks/python/apache_beam/typehints/typed_pipeline_test_py3.py]: reverted

File: sdks/python/apache_beam/transforms/core.py (hunk 1 above)
Comment (udim) [reply sdks/python/apache_beam/transforms/core.py]: I've reverted this case back to raise an exception.

**Comment by aaltay** [comment]
Could you look at the failing test?

**Comment by udim** [comment]
Ready for a final once-over. I had to rebase to get the latest lint checker running locally.
